File: LDXImagePicker/LDXAlbumTool.m
//
//  LDXAlbumTool.m
//  LDXImagePicker
//
//  Created by 刘东旭 on 4/9/25.
//  Copyright © 2025 LDX. All rights reserved.
//

#import "LDXAlbumTool.h"
#import "LDXImagePickerController.h"

@interface LDXAlbumBlock : NSObject <LDXImagePickerControllerDelegate>
@property (nonatomic, copy) void (^selectBlock)(NSArray<PHAsset *> *selectAssets);
@property (nonatomic, copy) void (^cancelBlock)(void);
@end

static NSMutableArray *_albumObjects;
@implementation LDXAlbumTool

+ (NSMutableArray *)albumObjects {
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        _albumObjects = [NSMutableArray new];
    });
    return _albumObjects;
}

+ (void)showAlbumWithController:(UIViewController *)vc
                   selectBlock:(void (^)(NSArray<PHAsset *> *))selectBlock
                   cancelBlock:(void (^)(void))cancelBlock
{
    LDXAlbumBlock *albumBlock = [LDXAlbumBlock new];
    [[self albumObjects] addObject:albumBlock];
    
    albumBlock.selectBlock = selectBlock;
    albumBlock.cancelBlock = cancelBlock;
    
    LDXImagePickerController *picker = [LDXImagePickerController new];
    picker.delegate = albumBlock;
    picker.mediaType = LDXImagePickerMediaTypeAny;
    picker.allowsMultipleSelection = YES;
    picker.showsNumberOfSelectedAssets = YES;
    picker.numberOfColumnsInPortrait = 3;
    picker.modalPresentationStyle = UIModalPresentationOverFullScreen;
    
    [vc presentViewController:picker animated:YES completion:nil];
}

@end

@implementation LDXAlbumBlock

#pragma mark - LDXImagePickerControllerDelegate
- (void)ldx_imagePickerControllerDidCancel:(LDXImagePickerController *)imagePickerController {
    [imagePickerController.presentingViewController dismissViewControllerAnimated:YES completion:nil];
    if (self.cancelBlock) self.cancelBlock();
    
    [[LDXAlbumTool albumObjects] removeObject:self];
}

- (void)ldx_imagePickerController:(LDXImagePickerController *)imagePickerController
             didFinishPickingAssets:(NSArray *)assets {
    [imagePickerController.presentingViewController dismissViewControllerAnimated:YES completion:nil];
    
    NSMutableArray<PHAsset *> *resultAssets = [NSMutableArray array];
    if ([assets isKindOfClass:[NSArray class]]) {
        for (id item in assets) {
            if ([item isKindOfClass:[PHAsset class]]) {
                [resultAssets addObject:item];
            }
        }
    }
    
    if (self.selectBlock) self.selectBlock(resultAssets);
    [[LDXAlbumTool albumObjects] removeObject:self];
}

@end

File: LDXImagePicker/LDXAlbumTool.swift
//
//  LDXAlbumTool.swift
//  TestDemo
//
//  Created by Mac-Mini on 2025/4/8.
//

import UIKit
import Photos
import LDXImagePicker

class LDXAlbumBlock: NSObject {
    var selectBlock: ((_ selectAssets: [PHAsset]) -> Void)?
    var cancelBlock: (() -> Void)?
}

extension LDXAlbumBlock: LDXImagePickerControllerDelegate {
    func ldx_imagePickerControllerDidCancel(_ imagePickerController: LDXImagePickerController!) {
        DispatchQueue.main.async {
            imagePickerController.dismiss(animated: true)
        }
        cancelBlock?()
        albumObjects.removeAll { $0 === self }
    }
    func ldx_imagePickerController(_ imagePickerController: LDXImagePickerController!, didFinishPickingAssets assets: [Any]!) {
        DispatchQueue.main.async {
            imagePickerController.dismiss(animated: true)
        }
        selectBlock?(assets as! [PHAsset])
        albumObjects.removeAll { $0 === self }
    }
}

nonisolated(unsafe) var albumObjects = [LDXAlbumBlock]()
public class LDXAlbumTool: NSObject {
    @MainActor static func showAlbum(vc: UIViewController, selectBlock: @escaping (_ selectAssets: [PHAsset]) -> Void, cancelBlock: @escaping (() -> Void)) {
        let albumBlock = LDXAlbumBlock()
        albumObjects.append(albumBlock)
        albumBlock.selectBlock = selectBlock
        albumBlock.cancelBlock = cancelBlock
        
        let picker = LDXImagePickerController()
        picker.delegate = albumBlock
        picker.mediaType = .any
        picker.allowsMultipleSelection = true
        picker.showsNumberOfSelectedAssets = true
        picker.numberOfColumnsInPortrait = 3
        picker.modalPresentationStyle = .overFullScreen
        vc.present(picker, animated: true)
    }

}
